src/animation.rs

```diff
@@ -45,11 +46,53 @@ impl Facing {
 #[derive(serde::Deserialize, Clone, Debug)]
 #[serde(deny_unknown_fields)]
 pub struct Clip {
+    #[serde(deserialize_with = "deserialize_range_from_array")]
     pub frames: Range<usize>,
     #[serde(default)]
     pub repeat: bool,
 }
 
+fn deserialize_range_from_array<'de, D>(de: D) -> Result<Range<usize>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    de.deserialize_tuple(2, RangeVisitor)
+}
+
+struct RangeVisitor;
+
+impl<'de> serde::de::Visitor<'de> for RangeVisitor {
+    type Value = Range<usize>;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("A sequence of 2 integers")
+    }
+
+    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+    where
+        A: SeqAccess<'de>,
+    {
+        let start: usize = if let Some(start) = seq.next_element()? {
+            start
+        } else {
+            return Err(serde::de::Error::invalid_length(
+                0,
+                &"a sequence with a length of 2",
+            ));
+        };
+        let end: usize = if let Some(end) = seq.next_element()? {
+            end
+        } else {
+            return Err(serde::de::Error::invalid_length(
+                1,
+                &"a sequence with a length of 2",
+            ));
+        };
+
+        Ok(start..end)
+    }
+}
+
```

Comment:
This reminds me of something I was wondering about before, which is not entirely relevant here, but I wonder if we can implement clips/animations which play a non-contiguous array of frame indexes. I had a usecase for it before when I was adding a waiting state and it didnt really match any of the available animation frames, but I'm not sure if that would be a good design. It would let us piece together frames from a spritesheet to make "ad-hoc" animations of the correct length, or skip frames in the middle of an animation if we needed to. The more I write it out the more janky it sounds though so maybe this is not a good idea :laughing:

Reply:
It could make sense. It would essentially be an array of ranges, which might not be bad.

Some of the exact details of the syntax become less important with an editor, too, so if it was a useful feature and we could make a good interface for it, it seems like it might be good to include.